Breaking: Version 2 improvements (fixes #29)

oliverfoster opened the pull request:
fixes #29

### Fix
* Reverted changes to broken intersection queries

### Update
* `isAvailableInHierarchy` to account for detached models
* sets sorted by `.order`
* removed the word raw throughout and changed for more accurate names, `availableComponents` etc
* removed the word sub throughout, from subset, and changed for more accurate names, `set`, `getSet..`, etc
* generally refactor code so that logic is more distinct and easier to read
* split scoringset into interactionset, lifecycleset and scoringset
* adaptmodelset extends interactionset,
* lifecycleset extends interactionset
* isolate the compatibility layer
* automatically assign set ids
* formalise `State` and `Objective` apis
* formalise lifecycle phases and triggers for `init`, `restore`, `start`, `reset`, `restart`, `visit`, `leave` and `update`, fully integrating [adapt-contrib-modifiers](https://github.com/adaptlearning/adapt-contrib-modifiers)
* add order for lifecycle elements so that [adapt-contrib-banking](https://github.com/adaptlearning/adapt-contrib-banking), [adapt-contrib-randomise](https://github.com/adaptlearning/adapt-contrib-randomise) and [adapt-contrib-scoringAssessment](https://github.com/adaptlearning/adapt-contrib-scoringAssessment) can execute in order, ordering sets appropriately
* separate utility functions into appropriate files 
* add jsdocs throughout

### New
* added multiple functions to account for detached models
* added `TotalSets` to replace the set behaviour on the main `Adapt.scoring` API
* document [INTERSECTION_QUERY.md](https://github.com/adaptlearning/adapt-contrib-scoring/blob/issue/29/INTERSECTION_QUERY.md)
* document [LIFECYCLE.md](https://github.com/adaptlearning/adapt-contrib-scoring/blob/issue/29/LIFECYCLE.md)

### Breaking
* Almost every function has been renamed or moved

oliverfoster committed:
Breaking: Version 2 improvements (fixes #29)

oliverfoster committed:
Markdown caps

oliverfoster committed:
Merdown edits

oliverfoster committed:
jsdoc updates

oliverfoster committed:
jsdoc fixes

oliverfoster committed:
jsdoc fixes

oliverfoster committed:
jsdoc updates

oliverfoster committed:
Formalise canReset

oliverfoster committed:
jsdoc typo

oliverfoster committed:
Extend adaptmodelset to include more questionmodel interfacing

oliverfoster committed:
Fixed findDescendantModels and getAllDescendantModels for detached models

joe-replin committed:
Resolving conflicts & Implementing logging improvements.

Logging improvements migrated from https://github.com/adaptlearning/adapt-contrib-scoring/pull/26

joe-replin committed:
Merge branch 'master' of https://github.com/adaptlearning/adapt-contrib-scoring into issue/29

joe-replin commented on js/helpers.js:
```diff
@@ -1,14 +1,19 @@
+    return sum(sets, 'score');
   }
 
   // score(context) {
```
Consider removal if no longer needed or comment if future plans.

oliverfoster replied:
I left in as Dan's unfinished work. @danielghost

joe-replin committed:
Schema nesting issue.

joe-replin committed:
Reverting FW version

danielghost commented on js/ScoringSet.js:
```diff
@@ -31,492 +27,220 @@ import _ from 'underscore';
-  get isObjectiveComplete() {
-    return this.isComplete;
+  get isSubmitted() {
+    return this.model.get('_isSubmitted');
```
A ScoringSet can be a collection of models. Should all logic targetting a single model be moved to the relevant ScoringSet? Currently we have a mixture within this super class.

oliverfoster replied:
The set lives on a model, its parent. Its children will be defined by itself. 
An extension to this class could change the isSubmitted to suit its own needs, but it probably shouldn't.
This is the basic underpinning of  selecting the AdaptModelSets. It is used in the query space as the 'isSubmitted' filter.

i.e. I don't understand your question. Could you give examples and reasons why you want to change it please?


> A ScoringSet can be a collection of models. 

Yes, its children.

> Should all logic targetting a single model be moved to the relevant ScoringSet? 

Why? This is returning the status of the parent here, by default, as a filter property. isSubmitted doesn't really work as a derived property of the children if you're filtering sets. It does work when you want to query submitted adapt models for any reason.

> Currently we have a mixture within this super class.

Where? Examples please.


Have a look at the query document https://github.com/adaptlearning/adapt-contrib-scoring/blob/issue/29/INTERSECTION_QUERY.md

danielghost replied:
A single `model` property wasn't part of `ScoringSet` in v1 as not everything is assigned to a single model, e.g. the concept of buckets which can be associated with many models but don't live on one. For cases such as this, we could leave the model `undefined`, or if a model was truly necessary, use the course model.

My main point was that within `ScoringSet` there is currently a mixture of aggregated calculations such as `score` and `correctness`, and then ones like `isSubmitted` and `isComplete` that target the single model. This could cause errors if a set didn't have a model and didn't override these methods. Most sets will likely have their own implementation for methods such as `isComplete`, but I would be tempted to either force a set to override these as we did in [v1](https://github.com/adaptlearning/adapt-contrib-scoring/blob/93c693963e3b94a6d8ec192b45a14c635f698605/js/ScoringSet.js#L470-L472) and currently used for `isPassed`, or make them derivable based on `availableModels` etc.

Unless unintentional, `AdaptModelSet` currently always returns the single model in an Array for `availableModels`, regardless of whether it has children, so the same `_isComplete` value would be calculated for both `this.model.get('_isComplete');` and `this.availableModels.every(model => model.get('_isComplete'));` anyway, with the latter having the added benefit of accounting for more potential use cases.

If `isSubmitted` is only really intended for `AdaptModelSet`, this could be moved into that class explicitly and still work with the query filters. I actually changed this class to extend `IntersectionSet` to prevent the scoring specific functionality that isn't required, such as logging score updates, creating objectives etc., so the class needs some changes regardless to work as expected. I would probably favour moving `isSubmitted` into that class. This filter property wasn't in v1, same with `isCorrect`, `isPartlyCorrect`, `isIncorrect` - is there a use case for needing to query any sets which aren't Adapt models by these?

oliverfoster replied:
You've moved the inheritance pattern around to accommodate the logging?

oliverfoster replied:
So, no scoring stuff or filters can now be used on the AdaptModelSet instances?

oliverfoster replied:
You want to:

1. move the Boolean filter properties to the AdaptModelSet only, except isComplete
3. remove the minScore maxScore stuff from AdaptModelSet by changing the inheritance 
4. extend the remaining filter functions, like isComplete, to do multiple children?

Have I got that right?

oliverfoster replied:
> A single model property wasn't part of ScoringSet in v1 as not everything is assigned to a single model, e.g. the concept of buckets which can be associated with many models but don't live on one. For cases such as this, we could leave the model undefined, or if a model was truly necessary, use the course model.

It represents the parent model for the set. Like randomise, banking and assessment all have a parent article, adaptmodelset has the model it represents etc. It allows us to update the sets in the lifecycle by doing parent model intersections and gives a good place around which children and configs and states can be derived.

A bucket would live at course level, (but it doesn't currently set this.model) and have children models from the course.

An assessment would live at article level but have blocks as children. Randomise and banking likewise.

oliverfoster replied:
> My main point was that within ScoringSet there is currently a mixture of aggregated calculations such as score and correctness, and then ones like isSubmitted and isComplete that target the single model. This could cause errors if a set didn't have a model and didn't override these methods. Most sets will likely have their own implementation for methods such as isComplete, but I would be tempted to either force a set to override these as we did in v1 and currently used for isPassed, or make them derivable based on availableModels etc.

I see, I think.

By moving the AdaptModelSet to IntersectionSet (~~for logging~~), one lower than ScoringSet, you've lost all of the filter properties and scoring properties on AdaptModelSet, and you now need to move the relevant ones to the AdaptModelSet

I did it this way round so that all of the filter properties applied to all of the sets all of the time and that I didn't need to divide them all over the different classes. They represent all of the main concepts in Adapt in one place like this. Even if they don't directly apply to every set all of the time. The idea was to make these main classes as broad as possible. It doesn't matter if isSubmitted returns undefined a lot of the time (probably need some optional changing for this.model?.), it isn't meaningful to isSubmitted an assessment, but it is meaningful to isSubmitted some of the AdaptModelSets. 

Because of the way intersections and filters work, if you wanted all of the `isSubmitted=false` or `isSubmitted=true` or `isSubmitted` (truthy) models, it will return only those where isSubmitted is not undefined, which would be only those where the concept was applicable i.e. Question components.

Reference https://github.com/adaptlearning/adapt-contrib-scoring/blob/c530c6dae06403e63cb8485bf87b8c646776e2d7/js/utils/intersection.js#L89
https://github.com/adaptlearning/adapt-contrib-scoring/blob/c530c6dae06403e63cb8485bf87b8c646776e2d7/js/ScoringSet.js#L237

I did try doing it your way before, and it got really messy really quickly, because you effectively need a second class for each adapt class. This way is simpler because it just works. You only use the properties which are meaningful to the concept you're working on. They either return a value or they don't. You can use the value to determine your filter. All Adapt concepts are in one place, completion, scoring, correctness, submitted etc.

It's a different way of thinking to the classic Adapt inheritance pattern, this is a broadly applicable query API that represents many abstract concepts pertinent to scoring. It returns undefined values where the API parts are not relevant to the course subject matter. Adapt has traditionally been a specific, applicable API oriented around the course content, which is why it's terrible for scoring, this is a scoring API oriented around all of the concepts in the ScoringSet.

May I suggest you put the inheritance back and sort out the logging?

danielghost replied:
No, the inheritance pattern wasn't switched just for logging. I thought it was actually a mistake for `AdaptModelSet` to extend `ScoringSet` as we don't want it to do all the things that scoring sets handle. It suggests this in the Class:
https://github.com/adaptlearning/adapt-contrib-scoring/blob/c530c6dae06403e63cb8485bf87b8c646776e2d7/js/AdaptModelSet.js#L4-L7

I hadn't considered the queries functionality when doing this though, as explained in the previous response, so I am aware this may need to be addressed. The set still works with intersection queries and when wanting to query the score related to a `ScoringSet` that intersects an `AdaptModelSet`, but not when the `AdaptModelSet` is the selector. I can set this back to extend from `ScoringSet` and then work out the best way to prevent `ScoringSet` functionality that we don't need in `AdaptModelSet`, rather than the other way around.

Regardless of the inheritance issue, there are points in my query that require a response to determine the base `ScoringSet` methods and calculations, as I still think it is currently inconsistent.

danielghost replied:
Apologies, I had my response above (re: https://github.com/adaptlearning/adapt-contrib-scoring/pull/30#discussion_r2966959108, https://github.com/adaptlearning/adapt-contrib-scoring/pull/30#discussion_r2966979516) waiting to be sent but got distracted and hadn't clicked the button, so only noticed your latest responses when the page updated after finally submitting. I will read these latest responses next week and it may make my latest comment redundant. Sorry in advance if this is the case.

oliverfoster replied:
> then work out the best way to prevent ScoringSet functionality that we don't need in AdaptModelSet

Could you describe what that functionality is and why it's causing you problems for it to be on the AdaptModelSet? The only functionality that I can see having been added is for the logging.

oliverfoster committed:
Fix wording in IntersectionSet description

Corrected wording for clarity in IntersectionSet section.

danielghost commented on js/ScoringSet.js:
```diff
@@ -1,27 +1,23 @@
+  sum
+} from './utils/math';
+import {
+  filterIntersectingHierarchy
```
This no longer exists. What is the replacement method?

Assume: https://github.com/adaptlearning/adapt-contrib-scoring/blob/8650fb544c83a08cf107d316a26bc67edd6d9f37/js/utils/models.js#L123

danielghost committed:
Update to remove the duplicated event trigger for 'scoring:update'. Moved event trigger for 'scoring:reset' to `Lifecycle` for consistency.

Questionable whether a public method to update scoring is required. No longer used directly in scoring plugins as was the case with v1.

danielghost committed:
Incorprated the log logic from https://github.com/adaptlearning/adapt-contrib-scoring/pull/26. Amended `AdaptModelSet` to extend `IntersectionSet` rather than `ScoringSet` as this is only required for intersection queries.

oliverfoster commented on js/ScoringSet.js:
```diff
@@ -57,7 +59,48 @@ export default class ScoringSet extends LifecycleSet {
+   * @param {Backbone.Model} model
+   * @returns {number}
+   */
+  getMinScoreByModel(model) {
```
If these functions, and those added by Joe, aren't used anywhere other than the logging code, can we move them out of the main ScoringSet API and into a helpers file? Or fold them into the places that they're called from? I don't see a need for multiple small functions when a single larger one would work.

This is almost 10 new main API functions, just for printing logs.

danielghost replied:
Will look into offloading the logic into another class, similar to how objectives are handled, but the different methods were setup to allow subclass extensibility, so we can easily override just the bits that differ between sets.

oliverfoster replied:
I don't see any use case for these beyond the logging stuff. Are you intending that the logging be overridden per set?

danielghost commented on js/ScoringSet.js:
```diff
@@ -217,306 +99,191 @@ export default class ScoringSet extends Backbone.Controller {
-    return this.models.some(model => model.get('_isVisited'));
+  get isPartlyCorrect() {
+    if (!this.isSubmitted) return null;
+    return (this.correctness < this.maxCorrectness);
```
Returns `true` if `correctness` is 0.

danielghost commented on js/ScoringSet.js:
```diff
@@ -217,306 +99,191 @@ export default class ScoringSet extends Backbone.Controller {
-    return this.isStarted && !this.isComplete;
+  get isIncorrect() {
+    if (!this.isSubmitted) return null;
+    return (!this.correctness && this.maxCorrectness);
```
Returns the `maxCorrectness` value if  `correctness` equals 0.

danielghost commented on js/Lifecycle.js:
```diff
@@ -0,0 +1,314 @@
+    },
+    // restart calls set.onStart when any intersecting model or set is reset
+    async restart(set) {
+      await set.onStart?.();
```
Could this call `set.onRestart?.()` instead to distinguish between a first attempt and a restart attempt? An alternative would be to have a flag akin to how `wasRestored` is used, but two different methods seems like a cleaner option. Is there a specific reason why this wouldn't work or cause problems?

See https://github.com/adaptlearning/adapt-contrib-scoring/pull/30/changes#r3024196814 for some context.

In cases where a set may require start and restart to be the same, `onRestart` called also call `onStart` within the set. Even in the current [`AssessmentSet.onStart`](https://github.com/adaptlearning/adapt-contrib-scoringAssessment/blob/7208b8b746bf619e3f8b7c9927029202e3c52f0c/js/AssessmentSet.js#L289)`, some of these methods probably only need to happen after a reset.

oliverfoster replied:
I didn't see any difference when I was rewriting everything. Just that reset/restart is usually exactly the same behaviour as start. Start is always the beginning of the lifecycle and reset, although seemingly at the end, is always the beginning of the lifecycle. I removed the on restart phase to clarify that it is effectively starting again. If you need to keep state between attempts then that state should be kept and managed on the set and not in the lifecycle stuff, where it's just confusing. Reset is not a finished or complete.

danielghost commented on js/ScoringSet.js:
```diff
@@ -606,97 +391,68 @@ export default class ScoringSet extends Backbone.Controller {
+    if (this.isIntersectedSet) return;
+    Logging.info(`${this.id} reset`);
+    super.reset();
+    this.objective?.reset();
```
The [score](https://github.com/adaptlearning/adapt-contrib-scoring/blob/c530c6dae06403e63cb8485bf87b8c646776e2d7/js/Objective.js#L38) isn't reset at this point as the set hasn't yet cleared its state. The state reset doesn't happen until after the functionality in `onStart` has executed.

oliverfoster replied:
This would give credance to an onRestart lifecycle handler.

danielghost commented on js/ScoringSet.js:
```diff
@@ -606,97 +391,68 @@ export default class ScoringSet extends Backbone.Controller {
-    this._setObjectiveStatus();
+  /** @override */
+  async onInit() {
+    if (this.isIntersectedSet) return;
```
Lifecycle methods such as `onInit`, `onRestore` etc. already have the intersected sets removed before the LifecycleRenderer calls them. Do we wish to keep the guards for any situation where these methods may be called outside of the renderer, or remove?

Sets which extend these methods, such as `AssessmentSet` don't have the guards, so they only partially protect the methods anyway. I would suggest the renderer is the enforcement point.

danielghost replied:
Leaving guards in as whilst they aren't needed, they show the intention that intersected sets shouldn't use these methods. Marked the appropriate lifecycle methods as `@protected` to indicate these shouldn't be called externally as that defeats the purpose of `LifecycleRenderer`.

danielghost committed:
Fixed `@typedef` paths and typos.

danielghost committed:
Reverted `AdaptModelSet` to extend `ScoringSet`. Amended filter methods in `ScoringSet` to query its models rather rather than the single parent model (which isn't always populated in a set). Fixed issues when evaluating correctness.

danielghost committed:
Adjusted lifecycle to distinguish between a first attempt and a restart attempt. Marked lifecycle callback hooks as protected methods, as calling these externally defeats the benefits of the LifecycleRenderer.

danielghost committed:
Updated objective functionality to incorporate https://github.com/adaptlearning/adapt-contrib-scoringAssessment/pull/17. Resolved reset issues as the objective needs to wait until the set has been reset before updating the score and status - adjusted methods and logic accordingly so score and status can be handled separately by the sets lifecycle methods.

danielghost committed:
Moved the log for the scoring updates functionality into a `Journal` class to keep this logic separate from the `ScoringSet`.

danielghost commented on js/AdaptModelSet.js:
```diff
@@ -1,97 +1,91 @@
-   * Intentionally empty to prevent super Class event triggers
-   * @override
-   */
-  onCompleted() {}
```
Do we need `AdaptModelSet` to be firing the scoring event triggers? Anything listening to events such as `'scoring:set:complete'` will be triggered every time one of these models is completed. Are we ever likely to use something like https://github.com/adaptlearning/adapt-contrib-scoringResults for these type of sets?

Leaving them in means we have a chatty scoring event system, which will probably not be utilised for these models. As these sets are only really needed to allow the queries and intersections to work, I would probably prevent them from firing the events to match v1, but happy to leave them in if it offers us more flexibility without compromising performance.

oliverfoster replied:
I think leave to them in; it's a valid event to fire as a set is being completed.

danielghost commented on js/adapt-contrib-scoring.js:
```diff
@@ -1,445 +1,218 @@
-  _updateQueue(model) {
-    this._queuedChanges.push(model);
-    this._debouncedUpdate();
+  getSetById(id) {
```
This is no longer exported. It can obviously be used via `Adapt.scoring.getSetById`, but should we make all of these types of methods utility functions that can be imported by plugins as required?

Currently we have a mixture of those that can only be referenced via `Adapt.scoring` and others which can be both imported individually or used via the scoring object. Do we need both, or just use imports?

oliverfoster replied:
They should be exported, yes.

`Adapt.scoring` references are easy to grab from the console. This can be handy for debugging.

danielghost replied:
Added exports. Kept references for console debugging.

danielghost commented on js/TotalSets.js:
```diff
@@ -0,0 +1,231 @@
+  /** @override */
+  onCompleted() {
+    if (this.isIntersectedSet) return;
+    Adapt.trigger('scoring:complete', Adapt.scoring);
```
Should use `this` as an argument rather than `Adapt.scoring`, to provide the correct context to listeners such as https://github.com/adaptlearning/adapt-contrib-core/blob/2c6ef07f8bfe5fe4dd317e4d0dad899cbd1f86e1/js/tracking.js#L93.

oliverfoster replied:
Ok

danielghost commented on js/Lifecycle.js:
```diff
@@ -0,0 +1,314 @@
+  async restore() {
+    const sets = getAllSets();
+    await this.renderer.render.restore(sets);
+    Adapt.trigger('scoring:restored', this.scoring);
```
As per https://github.com/adaptlearning/adapt-contrib-scoring/pull/30/changes#r3079115424, the event triggers on this file should include the `TotalSets` context unless we change the way the listeners currently work.

oliverfoster replied:
Ok

danielghost committed:
Fixed issue with `TotalSets` completing each time it was restored because the other sets hadn't yet been restored. Removed unnecessary overrides, with a decision to use inherited `ScoringSet` events and move from `'scoring:*'` to `'scoring:total:*'` - core framework listeners will be updated accoprdingly. Namespaced lifecycle events for clarity.

danielghost committed:
Renamed `Journal` to better reflect its purpose. Added a journal for `TotalSets` as the score logic for the modifiers is different, with the previous modifier scores not representing the score associated with that set.

danielghost committed:
Amended the journal for `TotalSets` to use its sets as the modifiers rather than the models, so it is easier to understand how the scores were updated. Fixed issue with logging modifier score changes when models become unavailable.

danielghost committed:
Exported some missing set utility functions.

danielghost committed:
Exported `Passmark` so this can be used by other plugins (could already but would need a direct import of the file, so exposed in same way as other methods and classes).

oliverfoster committed:
Journal refactor (#31)

danielghost committed:
Added guard checks to prevent false positives when checking `isComplete` and `isPassed`. Prevented potential for last registered set to be removed if attempting to deregister a set that hadn't been registered (no path in the current code where deregister receives an unregistered set, but added defensive coding in case a set authour breaks the inheritance chain).

danielghost committed:
Fixed backwards compatibility issue introduced with c2b06d5404a1d9e52ac3d4b169e14436bfc965b3.

danielghost commented on js/Lifecycle.js:
```diff
@@ -0,0 +1,315 @@
+   */
+  onScoringSetReset(set) {
+    // TODO: determine if we should restart the sets on this model id only or all descendant sets as well
+    if (!set.model) return;
```
Not all sets have a single model, so resetting one of these will not reset all intersected sets. Should any set without a parent model actually allow a reset anyway?

oliverfoster replied:
What happens when you give a parent model to your sets that currently have no parent model?

danielghost replied:
Adding a parent model to something like a BucketSet, which is configured on the course model, will cause all sets with this parent to reset, even if those sets models don't intersect any other models (excluding course) associated with the set being reset. This prevents the ability to only reset one set configured at the course level.

Suggest amending the logic to match the TODO comment and use:
```JavaScript
const sets = filterSetsByIntersectingModels(getAllSets(), set.models);
```

oliverfoster replied:
We should talk about this.

danielghost committed:
Amended passed/failed logic so this is only relevant for sets with a passmark criteria (fixes #28).

danielghost committed:
Prevent direct calls to `reset` when `canReset:false`. Fixed jsdoc issues in `Lifecycle`.

danielghost committed:
Updated the handebars helpers.

danielghost commented on js/helpers.js:
```diff
@@ -1,24 +1,75 @@
+   * @param {*} context
+   * @returns {number}
+   */
+  scaledScoreQuery(query, context) {
```
Do we need to distinguish between a `scaledScore` calculation from the score values, and a sum of the `scaledScore` from each returned set? These can be very different values depending on how the scores were assigned. For example, 3 sets consisting of 4 question models with the following scoring breakdown:

| id | minScores | maxScores | scores | scaledScore |
| --- | --- | --- | --- | --- |
| 1 | 0,0,0,0 | 1,1,1,1 | 0,1,1,1 | 75 |
| 2 | 0,0,0,0 | 10,10,10,10 | 10,0,0,10 | 50 | 
| 3 | 0,0,0,0 | 100,100,100,100 | 0,100,0,0 | 25 |

Creating a new scaledScore using `getScaledScoreFromMinMax` will return **28**: ((3 + 20 + 100) /  (4 + 40 + 400)) x 100.

Summing all the scaledScores will return **50**: ((75 + 50 + 25) / (100 + 100 + 100)) x 100;

The current `getScaledScoreFromMinMax` logic assumes the scoring bias/weighting has been considered when distributing the scores. Summing the already calculated `scaledScore` for each set, treats the performance of each set the same.

Should the overall score be computed by rescaling raw scores across all questions (point-weighted), or by aggregating already-scaled set scores (set-weighted)?

oliverfoster replied:
tldr; we don't have a design pattern for what the total course scaled score means, if it should be the `average assessment scaled score` or the `average item scaled score`. These equate to __"Did the user do well on all assessments?"__ or  __"Did the user do well on the most important items?"__ . `average item scaled score` appears more useful as it carries any present item weighting through to the final course scaled score.

### Discovery

| Aspect | Score | ScaledScore | Weighting |
| --- | --- | --- | --- | 
| item | score | score / maxScore  | n/a |
| set | sum of item scores | sum of item scores / sum of item maxScores |  relative size of item max scores |
| sets | sum of set scores | sum of set scores / sum of set maxScores | relative size of item max scores |


| set id | item minScores | item maxScores | item scores | set scaledScore |
| --- | --- | --- | --- | --- |
| 1 | 0,0,0,0 | 1,1,1,1 | 0,1,1,1 | (3/4) = 75% |
| 2 | 0,0,0,0 | 10,10,10,10 | 10,0,0,10 | (20/40) = 50% | 
| 3 | 0,0,0,0 | 100,100,100,100 | 0,100,0,0 | (100/400) = 25% |

The `average assessment scaled score` is (75%+50%+25%) / 3 = 50%. Which is how the assessment vs. total assessment works in Adapt currently. We weigh the individual assessments relative to each other. It works nicely if you have some sections with low point values and some with high point values, where you want to determine the `average assessment scaled score` across the course. __"Did the user do well on all assessments?"__

If you want to know the `average item scaled score`, which we currently don't do, you could sum the collective maxScore and score, either via total or average:

| | total minScores | total maxScores | total scores | total items scaledScore |
| --- | --- | --- | --- | --- |
| sets totals | 0 | (4+10+400) = 444 | (3+20+100) = 123 | (123/444) = ~28% |
| sets avgs | 0 | (4+10+400) / 3 = 148  | (3+20+100) / 3 = 41 | (41/148) = ~28% |

That would be good for __"Did the user do well on the most important items?"__

### Conclusion

To turn this on its head, if all items were weighted equally, both values are equal.

| set id | item minScores | item maxScores | item scores | set scaledScore |
| --- | --- | --- | --- | --- |
| 1 | 0,0,0,0 | 1,1,1,1 | 0,1,1,1 | (3/4) = 75% |
| 2 | 0,0,0,0 | 1,1,1,1 | 1,0,0,1 | (2/4) = 50% | 
| 3 | 0,0,0,0 | 1,1,1,1 | 0,1,0,0 | (1/4) = 25% |

`average assessment scaled score` = 50%

| | total minScores | total maxScores | total scores | total items scaledScore |
| --- | --- | --- | --- | --- |
| sets totals | 0 | (4+4+4) = 12 | (3+2+1) = 6 | (6/12) = 50% |
| sets avgs | 0 | (4+4+4) / 3 = 4  | (3+2+1) / 3 = 2 | (2/4) = 50% |

`average item scaled score` = 50%

Which would suggest that `average item scaled score` has the advantage because it reflects item weighting if any weighting is present.

danielghost replied:
Added `averageScaledScore` logic. `isPassed` still uses `scaledScore` (point-weighted/average item scaled score), but the methods required to return an average set scaled score are now included. See 71a5dabb399f08407cdaee36e0af262b39f9136e.

danielghost committed:
Added `averageScaledScore` to allow a distinction between point-weighted and set-weighted scaled scores. See https://github.com/adaptlearning/adapt-contrib-scoring/pull/30/changes#r3129702139.

danielghost committed:
Amended the lifecycle so it correctly resets sets based on the intersecting models of the set being reset - see https://github.com/adaptlearning/adapt-contrib-scoring/pull/30/changes#r3173433396. Tidied some jsdoc comments.

danielghost committed:
Fixed scope issues with the compatibility state.
